Admin Page: Move to Core translation functions -- PART 1

akirk opened the pull request:
#### Changes proposed in this Pull Request:

**This will need to happen in steps so reviews are easier. This is step 1.**

We're aiming to move away from the `i18n-calypso` tooling to handle translations. Instead we'll rely on 2 things:
- `@wordpress/i18n` for most translation work: `__`, `_n`, `_x`, and `sprintf`.
- `@wordpress/element` and its `createInterpolateElement` to handle React elements in strings.

`i18n-calypso` will remain in use for a few exceptions for now:
- `moment`
- `numberFormat`

This first PR introduces those changes for the At A Glance dashboard.

#### Jetpack product discussion

* p1HpG7-99c-p2
* Primary issue: #16481 

#### Does this pull request change what data or activity we track or use?

* N/A

#### Testing instructions:

Nothing should change at first sight:

- Under Jetpack > Dashboard, check how things look like everywhere on the page when you've connected your site to WordPress.com, or when Development mode is active.
- Try with different plugins and module active / inactive to get different card statues.
- Try with outdated plugins to see different messages.
- If you view source on the page, you should see a new section right before the React JS bundle is loaded: `react-plugin-js-translations`. There are no translations here for now, as no language pack is generated yet for this JS bundle. We'll need to push an Alpha first.

#### Proposed changelog entry for your changes:

* TBD


******************************

#### Original PR description below (before the discussion in the comments below happened)

This is a demo PR that shows how we'll be able to remove all of our uses of `translate()` (from i18n-calypso) within the Jetpack Dashboard so that we can move over to using language pack JSONs for this.

This could be merged as is and would have the Akismet "At a glance" use translations loaded the standard WordPress/Gutenberg way (i.e. using `wp_set_script_translations()`). The downside of this is that this duplicates the translations JSON in the HTML source.

To bring this over the finish line we'll also need to change our build process so that the translatable strings are picked up from the JavaScript source. We currently only ship the uglified `_inc/build/admin.js` and not the JavaScript clean source. If we committed the unmodified source to SVN, the strings would be extracted for translation and provided inside language packs.

To avoid duplication, the [convention in building the language packs](https://meta.trac.wordpress.org/browser/sites/trunk/wordpress.org/public_html/wp-content/plugins/wporg-gp-customizations/inc/cli/class-language-pack.php#L441) is to have the source be placed under a `src/` path, so that they won't be included in the language pack.

akirk committed:
Convert DashAkismet to use wordpress/i18n with interpolate-components

akirk commented:
> I think minified JS is fine; wp i18n was smart enough to pick strings from the bundles produced from gutenberg-extensions (built to _inc/blocks). You could confirm locally running the CLI command?

Nice, I just verified that my translatable strings appear with a JS source in the POT file after running `wp i18n make-pot . test.pot`:
```
#: _inc/jetpack-strings.php:75
#: _inc/build/admin.js:543
msgid "Already have a key? {{a}}Activate Akismet{{/a}}"
msgstr ""
```

The reason I assumed we'd need to ship the true source was because I searched for other usages of `__()` from `@wordpress/i18n` in the source and, for example, found this: 
https://github.com/Automattic/jetpack/blob/4b712c79f7c61442086f5801fcc16dfa16bd48a0/modules/search/instant-search/components/search-results.jsx#L6
https://github.com/Automattic/jetpack/blob/4b712c79f7c61442086f5801fcc16dfa16bd48a0/modules/search/instant-search/components/search-results.jsx#L19

This string is not translatable: https://translate.wordpress.org/projects/wp-plugins/jetpack/dev/de/default/?filters%5Bterm%5D=You+are+searching+for&filters%5Buser_login%5D=&filters%5Bstatus%5D=current_or_waiting_or_fuzzy_or_untranslated&filter=Filter&sort%5Bby%5D=priority&sort%5Bhow%5D=desc

The reason for why this was not picked up is the missing `jetpack` textdomain, if I change it to
```
<p>{ sprintf( __( 'You are searching for: "%s"', 'jetpack' ), query ) }</p>
```
It's picked up properly. So we need to keep this in mind when converting.

akirk commented:
I think it'd be interesting to write a code mod for this?

simison commented:
Noting that https://github.com/WordPress/gutenberg/pull/16374 uses `<`s instead of `{{`s and seems actually pretty close being merged as experimental API.

zinigor commented:
Let's look at this for the next milestone.

simison commented:
This could now be updated to use `__experimentalCreateInterpolateElement` available in `@wordpress/i18n` (https://github.com/WordPress/gutenberg/pull/17376) since Gutenberg 7.0

simison commented:
`@wordpress/i18n` now has stable `createInterpolateElement`.

WP 5.4 is shipped too — time to move to `wp-i18n` fully? :-)

ocean90 commented:
> We currently only ship the uglified _inc/build/admin.js and not the JavaScript clean source. If we committed the unmodified source to SVN, the strings would be extracted for translation and provided inside language packs.

Worth to mention that [source maps](https://github.com/wp-cli/i18n-command/blob/master/src/MapCodeExtractor.php) are also supported. This was [recently tested with BuddyPress](https://buddypress.trac.wordpress.org/ticket/8290).

jeherve committed:
Switch from an external dependency to @wordpress/i18n

@wordpress/i18n now ships with `createInterpolateElement`, that should be just what we need here.

jeherve commented on _inc/client/at-a-glance/akismet.jsx:
```diff
@@ -4,7 +4,9 @@
 import React, { Component } from 'react';
 import { connect } from 'react-redux';
-import { numberFormat, translate as __ } from 'i18n-calypso';
+import { numberFormat } from 'i18n-calypso';
```
Let's aim to address that in a follow-up PR. I noted this in #16481.

jeherve committed:
Update all At A Glance files to use @wordpress tooling

jeherve commented on _inc/client/at-a-glance/plugins.jsx:
```diff
@@ -70,26 +72,22 @@ class DashPluginUpdates extends Component {
+				{ updatesAvailable && <h2 className="jp-dash-item__count">{ pluginUpdates.count }</h2> }
 				<p className="jp-dash-item__description">
 					{ updatesAvailable
 						? [
-								__( 'Plugin needs updating.', 'Plugins need updating.', {
-									count: pluginUpdates.count,
-								} ) + ' ',
+								_n(
+									'Plugin needs updating.',
+									'Plugins need updating.',
+									pluginUpdates.count,
+									'jetpack'
+								) + ' ',
 								! this.props.isDevMode &&
-									__( '{{a}}Turn on plugin autoupdates.{{/a}}', {
-										components: { a: <a href={ managePluginsUrl } /> },
+									createInterpolateElement( __( '<a>Turn on plugin autoupdates.</a>', 'jetpack' ), {
+										a: <a href={ managePluginsUrl } />,
```
We're going to need to update this in a future PR. I don't think this is easy to translate, the number should be part of the string to translate.

jeherve commented on _inc/client/at-a-glance/scan.jsx:
```diff
@@ -188,25 +192,25 @@ class DashScan extends Component {
 					<h2 className="jp-dash-item__count is-alert">{ numberFormat( numberOfThreats ) }</h2>,
 					<p className="jp-dash-item__description">
-						{ __(
-							'Security threat found. Please {{a}}fix it{{/a}} as soon as possible.',
-							'Security threats found. Please {{a}}fix these{{/a}} as soon as possible.',
+						{ createInterpolateElement(
+							_n(
+								'Security threat found. Please <a>fix it</a> as soon as possible.',
+								'Security threats found. Please <a>fix these</a> as soon as possible.',
+								numberOfThreats,
+								'jetpack'
+							),
```
Same as https://github.com/Automattic/jetpack/pull/13527/files#r454473794

jeherve commented on _inc/lib/admin-pages/class.jetpack-react-page.php:
```diff
@@ -180,6 +180,8 @@ function page_admin_scripts() {
 			wp_enqueue_script( 'jp-tracks', '//stats.wp.com/w.js', array(), gmdate( 'YW' ), true );
 		}
 
+		wp_set_script_translations( 'react-plugin', 'jetpack', JETPACK__PLUGIN_DIR . '/languages/json/' );
```
@akirk I don't quite understand `JETPACK__PLUGIN_DIR . '/languages/json/'`. Do the translations need to live within the Jetpack plugin? Isn't the default good enough? Is this because we want to use the translations we pull manually on each release instead of the ones provided by language packs? If we do that, does this mean that folks will need to wait to get new strings until each new release?

akirk replied:
I might have written this code at a time where this was not an option yet. Clearly the path should be removed if we use language packs.

jeherve replied:
Ok, thanks! Done in 56c92cba1c77f4018eb402b253c59a630fcf1c6c

jeherve committed:
Rely on language packs instead

jeherve commented:
Here is how the updated strings look like on the production build:

`:Object(xh.__)("Activity","jetpack")`

`(Object(xh.__)("Already have a key? <a>Activate Akismet</a>","jetpack")`

`Object(xh._x)("Spam comments blocked.",'Example: "412 Spam comments blocked"',"jetpack")`

I believe translate.wordpress.org should be able to pick strings from this.

ocean90 commented:
> I believe translate.wordpress.org should be able to pick strings from this.

You can test this by running the [make-pot command from WP-CLI](https://developer.wordpress.org/cli/commands/i18n/make-pot/) since that's what the importer is using too.

kraftbj commented:
I'm getting a fatal JS error at wp-admin/admin.php?page=jetpack#/dashboard but I'm still trying to track down why.

```
Uncaught TypeError: Object(...) is not a function
    value https://historic-termite.jurassic.ninja/wp-content/plugins/jetpack-dev/_inc/build/admin.js?ver=8.8-alpha:71
    value https://historic-termite.jurassic.ninja/wp-content/plugins/jetpack-dev/_inc/build/admin.js?ver=8.8-alpha:71
    finishClassComponent https://historic-termite.jurassic.ninja/wp-includes/js/dist/vendor/react-dom.js?ver=16.9.0:17177
    updateClassComponent https://historic-termite.jurassic.ninja/wp-includes/js/dist/vendor/react-dom.js?ver=16.9.0:17132
    beginWork$1 https://historic-termite.jurassic.ninja/wp-includes/js/dist/vendor/react-dom.js?ver=16.9.0:18643
    callCallback https://historic-termite.jurassic.ninja/wp-includes/js/dist/vendor/react-dom.js?ver=16.9.0:341
    invokeGuardedCallbackDev https://historic-termite.jurassic.ninja/wp-includes/js/dist/vendor/react-dom.js?ver=16.9.0:391
    invokeGuardedCallback https://historic-termite.jurassic.ninja/wp-includes/js/dist/vendor/react-dom.js?ver=16.9.0:448
    beginWork$$1 https://historic-termite.jurassic.ninja/wp-includes/js/dist/vendor/react-dom.js?ver=16.9.0:23355
....
```

Steps to reproduce:
* New site.
* Set language to es_MX (Español de México).
* Install JP if not installed. Update translations.
* Connect JP.
* Go to the Dashboard.

ocean90 commented:
@kraftbj Probably because `wp.element.createInterpolateElement` only exists as `wp.element.__experimentalCreateInterpolateElement` in WordPress 5.4. It will become stable in 5.5 as `wp.element.createInterpolateElement`.

jeherve commented:
> I'm getting a fatal JS error at wp-admin/admin.php?page=jetpack#/dashboard

That will teach me for using trunk on my dev site. My bad. 🤦 

> Probably because wp.element.createInterpolateElement only exists as wp.element.__experimentalCreateInterpolateElement in WordPress 5.4. It will become stable in 5.5 as wp.element.createInterpolateElement.

That's it indeed. In #16478, I was aiming to use the standalone `@wordpress/element` package (version 2.16.0, which includes `createInterpolateElement`) to build the dashboard, instead of relying on the packages provided in Core. I had forgotten that a while ago we switched to externalizing the core dependencies from to lower the size of the dashboard bundle. I'll work on a workaround.

jeherve committed:
Admin Page: create new create interpolate element

This element acts as a wrapper for Core's own createInterpolateElement.

createInterpolateElement is available in WordPress 5.5 and in latest versions of the Gutenberg plugin, but it is not available in WordPress 5.4, which we still support.
This new component can consequently be used for now, until WordPress 5.5 becomes the minimum required version for Jetpack. When that happens we can deprecate this component and rely on createInterpolateElement directly.

jeherve committed:
Switch to using new jetpackCreateInterpolateElement component

See 7c6e0dd00cf9a7ef2cd8b51ed5a6930cb6141b48
